### PR TITLE
[AutoDiff] Remap `apply` callee type in derivative context.

### PR DESCRIPTION
**Pull request by dan-zheng**
Previously, `LinearMapInfo::addLinearMapToStruct` did not remap `apply` callee
type in derivative context.

Now, remapping is done. Remapping is significant when the derivative has a
more constrained generic signature.

Resolves [TF-817](https://bugs.swift.org/browse/TF-817).

---

Reproducer:
```swift
extension Array {
  @differentiable(wrt: self where Element : Differentiable)
  public func foo(index: Int) -> Element {
    return self[0]
  }
}
```

```
# 'self` wrt parameter type.

# Before: 
(bound_generic_struct_type decl=Swift.(file).Array
  (primary_archetype_type address=0x7fba3e0e1990 name=Element
))

# After:
(bound_generic_struct_type decl=Swift.(file).Array
  (primary_archetype_type address=0x7fba3d95e008 conforms_to=Swift.(file).Differentiable name=τ_0_0

    (nested_type=TangentVector <<unresolved>>)))
```

### CODE REVIEW

**Comment by dan-zheng** [comment]
Note: the actual fix is relatively small (https://github.com/apple/swift/pull/27590/commits/9fffe8960b92c1630927bd3103858ab387068a67).
I also included NFC gardening (https://github.com/apple/swift/pull/27590/commits/0501f5ffa4d8325a3391445557a22be71fb6f443) to save time.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 1]
```diff
@@ -4243,11 +4257,11 @@ class JVPEmitter final
   }
 
   static SILBuilder
-  initializeDifferentialAndBuilder(ADContext &context, SILFunction *original,
-                                   SILDifferentiableAttr *attr,
-                                   LinearMapInfo *linearMapInfo) {
+  initializeDifferentialBuilder(ADContext &context, SILFunction *original,
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: I'm very surprised this function existed. It's pretty specific to differentials and has only 2 lines. Inline this?

**Review by rxwei — APPROVED** [review]

File: lib/SILOptimizer/Mandatory/Differentiation.cpp (hunk 1 above)
Comment (dan-zheng) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: Done in https://github.com/apple/swift/pull/27590/commits/b2481dee4ded16998f0d03dfdf216f66557813c1.

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by rxwei** [comment]
@swift-ci Please test tensorflow

**Pull request merged**